src/NSrtm.Core/Pgm/PgmElevationProvider.cs

```diff
@@ -0,0 +1,119 @@
+﻿using System;
+using System.Collections.Concurrent;
+using System.Collections.Generic;
+using System.Linq;
+using System.Threading.Tasks;
+using JetBrains.Annotations;
+using NSrtm.Core.BicubicInterpolation;
+using NSrtm.Core.Pgm.DataDesciption;
+using NSrtm.Core.Pgm.GeoidUndulationGrid;
+
+namespace NSrtm.Core.Pgm
+{
+    internal sealed class PgmElevationProvider : IElevationProvider
+    {
+        private readonly IPgmGeoidUndulationGrid _discreteSurface;
+
+        private readonly ConcurrentDictionary<PgmCellCoords, BivariatePolynomial> _continuousSurface =
+            new ConcurrentDictionary<PgmCellCoords, BivariatePolynomial>();
+
+        public PgmElevationProvider([NotNull] IPgmGeoidUndulationGrid discreteSurface)
+        {
+            if (discreteSurface == null) throw new ArgumentNullException("discreteSurface");
+            _discreteSurface = discreteSurface;
+        }
+
+        private PgmDataDescription dataDescription { get { return _discreteSurface.PgmParameters; } }
+
+        private BivariatePolynomial getInterpolationForCellSurface(PgmCellCoords pgmCellCoords)
+        {
+            var nodesCoordinates = findCellAndSurroundingNodesCoords(pgmCellCoords,
+                                                                     dataDescription.LatitudeIncrementDegrees,
+                                                                     dataDescription.LongitudeIncrementDegrees);
+            var nodesUndulations = nodesCoordinates.Select(coor => _discreteSurface.GetUndulation(coor.Lat, coor.Lon));
+            var formattedUndulations = nodesUndulations.Select((c, i) => new {Index = i, value = c})
+                                                       .GroupBy(p => p.Index / 4)
+                                                       .Select(c => c.Select(v => v.value)
+                                                                     .ToList())
+                                                       .ToList();
+            return BicubicCalculator.GetSpline(values: formattedUndulations, step: dataDescription.LatitudeIncrementDegrees);
+        }
+
+        private static PgmCellCoords normalizeCoords(double lat, double lon)
+        {
+            var normalizedLongitude = lon;
+            if (lat > 90 || lat < -90)
+            {
+                normalizedLongitude += 180;
+            }
+            normalizedLongitude = (360 + (normalizedLongitude % 360)) % 360;
+            var absoluteLat = Math.Abs(Math.Abs(lat % 180) - 90);
```

Comment:
Maybe you can simplify this. If you want read how to ganerate triangle wave i recommend this article http://www.hitmaroc.net/38786-5090-there-one-line-function-that-generates-triangle-wave.html